conn.go

```diff
@@ -402,218 +387,160 @@ func (c *conn) DeleteSession(s *Session) {
 	c.channels.Remove(uint32(s.channel))
 }
 
-// mux is started in it's own goroutine after initial connection establishment.
-// It handles muxing of sessions, keepalives, and connection errors.
-func (c *conn) mux() {
-	var (
-		// map channels to sessions
-		sessionsByRemoteChannel = make(map[uint16]*Session)
-	)
-
-	// hold the errMu lock until error or done
-	c.errMu.Lock()
-	defer c.errMu.Unlock()
-	defer c.close() // defer order is important. c.errMu unlock indicates that connection is finally complete
+// connReader reads from the net.Conn, decodes frames, and either handles
+// them here as appropriate or sends them to the session.rx channel.
+func (c *conn) connReader() {
+	defer func() {
+		close(c.rxDone)
+		c.closeOnce.Do(func() { c.close() })
+	}()
 
+	var sessionsByRemoteChannel = make(map[uint16]*Session)
+	var err error
 	for {
-		// check if last loop returned an error
-		if c.err != nil {
+		if err != nil {
+			debug.Log(1, "connReader terminal error: %v", err)
+			c.rxErr = err
 			return
 		}
 
-		select {
-		// error from connReader
-		case c.err = <-c.connErr:
-
-		// new frame from connReader
-		case fr := <-c.rxFrame:
-			var (
-				session *Session
-				ok      bool
-			)
-
-			switch body := fr.Body.(type) {
-			// Server initiated close.
-			case *frames.PerformClose:
-				if body.Error != nil {
-					c.err = body.Error
-				}
-				return
-
-			// RemoteChannel should be used when frame is Begin
-			case *frames.PerformBegin:
-				if body.RemoteChannel == nil {
-					// since we only support remotely-initiated sessions, this is an error
-					// TODO: it would be ideal to not have this kill the connection
-					c.err = fmt.Errorf("%T: nil RemoteChannel", fr.Body)
-					break
-				}
-				c.sessionsByChannelMu.RLock()
-				session, ok = c.sessionsByChannel[*body.RemoteChannel]
-				c.sessionsByChannelMu.RUnlock()
-				if !ok {
-					c.err = fmt.Errorf("unexpected remote channel number %d", *body.RemoteChannel)
-					break
-				}
+		if c.idleTimeout > 0 {
+			_ = c.net.SetReadDeadline(time.Now().Add(c.idleTimeout))
+		}
 
-				session.remoteChannel = fr.Channel
-				sessionsByRemoteChannel[fr.Channel] = session
+		var fr frames.Frame
+		fr, err = c.readFrame()
```

Comment:
NIT: f, err := c.readFrame()?

Reply:
This is due to `err` being declared outside the loop and we don't want it to be shadowed.